### PR TITLE
Fraud Proofs - one PR to rule them all

### PR DESCRIPTION
**Pull request by tzdybal**
This PR wraps up entire Fraud Proof related work into single PR. It might be easier to review / apply comments this way.

Replaces: #375, #498, #500.

### CODE REVIEW

File: node/integration_test.go [hunk 1]
```diff
@@ -159,7 +160,88 @@ func TestTxGossipingAndAggregation(t *testing.T) {
...
+			nodeBlock, err := nodes[i].Store.LoadBlock(h)
+			require.NoError(err)
+			aggBlock, err := nodes[0].Store.LoadBlock(h)
+			require.NoError(err)
+			assert.Equal(aggBlock, nodeBlock)
+		}
+	}
+}
+
+func createNodes(isMalicious bool, num int, wg *sync.WaitGroup, t *testing.T) ([]*Node, []*mocks.Application) {
```
Comment (tzdybal) [inline node/integration_test.go:244]:
```suggestion
func createNodes(num int, isMalicious bool, wg *sync.WaitGroup, t *testing.T) ([]*Node, []*mocks.Application) {
```

File: proto/tendermint/abci/types.proto [hunk 2]
```diff
@@ -212,6 +233,12 @@ message ResponseCheckTx {
   repeated Event events     = 7
       [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
   string codespace = 8;
+  string sender    = 9;
+  int64  priority  = 10;
+
+  // mempool_error is set by Tendermint.
+  // ABCI applictions creating a ResponseCheckTX should not set mempool_error.
+  string mempool_error = 11;
```
Comment (tzdybal) [inline proto/tendermint/abci/types.proto:241]: Is this update from upstream Tendermint?

File: proto/tendermint/version/types.proto [hunk 3]
```diff
@@ -0,0 +1,24 @@
+syntax = "proto3";
```
Comment (tzdybal) [inline proto/tendermint/version/types.proto:1]: Is this file needed? Or it's just side-effect of running `./proto/get_deps.sh`?

File: proto/tendermint/abci/types.proto (hunk 2 above)
Comment (Manav-Aggarwal) [reply proto/tendermint/abci/types.proto:241]: yes, not from me

File: proto/tendermint/abci/types.proto [hunk 4]
```diff
@@ -402,6 +468,11 @@ service ABCIApplication {
   rpc EndBlock(RequestEndBlock) returns (ResponseEndBlock);
   rpc ListSnapshots(RequestListSnapshots) returns (ResponseListSnapshots);
   rpc OfferSnapshot(RequestOfferSnapshot) returns (ResponseOfferSnapshot);
-  rpc LoadSnapshotChunk(RequestLoadSnapshotChunk) returns (ResponseLoadSnapshotChunk);
-  rpc ApplySnapshotChunk(RequestApplySnapshotChunk) returns (ResponseApplySnapshotChunk);
+  rpc LoadSnapshotChunk(RequestLoadSnapshotChunk)
+      returns (ResponseLoadSnapshotChunk);
+  rpc ApplySnapshotChunk(RequestApplySnapshotChunk)
+      returns (ResponseApplySnapshotChunk);
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply proto/tendermint/abci/types.proto:474]: yes, not from me

File: proto/tendermint/version/types.proto (hunk 3 above)
Comment (Manav-Aggarwal) [reply proto/tendermint/version/types.proto:1]: it's just side-effect of running ./proto/get_deps.sh

File: block/manager.go [hunk 5]
```diff
@@ -228,6 +230,22 @@ func (m *Manager) SyncLoop(ctx context.Context) {
 			if err != nil {
 				m.logger.Info("failed to sync next block", "error", err)
 			}
+		case fraudProof := <-m.FraudProofCh:
+			m.logger.Debug("fraud proof received", "Block Height", "dummy block height") // TODO: Insert real block height
+			_ = fraudProof
+			// TODO(light-client): Set up a new cosmos-sdk app
+			// How to get expected appHash here?
```
Comment (liamsi) [inline block/manager.go:237]: Is this tracked somewhere?

File: node/integration_test.go [hunk 6]
```diff
@@ -166,7 +167,91 @@ func TestTxGossipingAndAggregation(t *testing.T) {
...
+	aggCtx, aggCancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithCancel(context.Background())
+	nodes, apps := createNodes(aggCtx, ctx, clientNodes+1, true, &wg, t)
+
+	wg.Add((clientNodes + 1) * clientNodes)
+	for _, n := range nodes {
+		require.NoError(n.Start())
+	}
+
+	time.Sleep(1 * time.Second)
```
Comment (liamsi) [inline node/integration_test.go:185]: Document what this sleep is for (and ideally why one second).

File: node/node.go [hunk 7]
```diff
@@ -349,6 +350,23 @@ func (n *Node) newCommitValidator() p2p.GossipValidator {
...
+func (n *Node) newFraudProofValidator() p2p.GossipValidator {
+	return func(fraudProofMsg *p2p.GossipMessage) bool {
+		n.Logger.Debug("fraud proof received", "from", fraudProofMsg.From, "bytes", len(fraudProofMsg.Data))
+		var fraudProof types.FraudProof
+		err := fraudProof.UnmarshalBinary(fraudProofMsg.Data)
+		if err != nil {
+			n.Logger.Error("failed to deserialize fraud proof", "error", err)
+			return false
+		}
+		// TODO(manav): Add validation checks for fraud proof here
```
Comment (liamsi) [inline node/node.go:364]: This should be tracked somewhere else than just in a todo.

File: proto/tendermint/abci/types.proto (hunk 2 above)
Comment (liamsi) [reply proto/tendermint/abci/types.proto:241]: Is there a way to get the changes from upstream tendermint into a separate PR. Will reduce the diff and make this PR easier to review.

File: proto/tendermint/abci/types.proto (hunk 4 above)
[the remark this answers is not in the document]
Comment (liamsi) [reply proto/tendermint/abci/types.proto:474]: Should not be part of this PR imo.

File: state/executor.go [hunk 8]
```diff
@@ -288,35 +317,132 @@ func (e *BlockExecutor) execute(ctx context.Context, state types.State, block *t
...
+			return nil, err
+		}
+		ISRs = append(ISRs, isr)
+		isFraud := e.checkFraudProofTrigger(isr, currentIsrs, currentIsrIndex)
+		if isFraud {
+			fraudProof, err := e.generateFraudProof(&beginBlockRequest, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			// TODO: gossip fraudProof to P2P network
```
[the remark this answers is not in the document]
Comment (liamsi) [reply state/executor.go:355]: All TODOs should be tracked in the/a fraud proof tracking issue!

File: types/fraudproof.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	abci "github.com/tendermint/tendermint/abci/types"
+	tmcrypto "github.com/tendermint/tendermint/proto/tendermint/crypto"
+)
+
+// Represents a single-round fraudProof
+type FraudProof struct {
+	// The block height to load state of
+	blockHeight int64
+
+	appHash []byte
```
Comment (liamsi) [inline types/fraudproof.go:13]: I think appHash should be documented as well.

File: block/manager.go [hunk 10]
```diff
@@ -228,6 +230,22 @@ func (m *Manager) SyncLoop(ctx context.Context) {
...
+
+			// success, err := m.executor.VerifyFraudProof(fraudProof, nil)
```
Comment (liamsi) [inline block/manager.go:239]: Isn't this code essential for this PR?

Comment (Manav-Aggarwal) [reply block/manager.go:239]: This isn't possible currently because we don't have a fraud proof window implemented. Made a tracking issue for it [here](https://github.com/celestiaorg/rollmint/issues/549).

File: types/fraudproof.go (hunk 9 above)
Comment (Manav-Aggarwal) [reply types/fraudproof.go]: Currently removed this updated structure of Fraud Proof in order to avoid adding serialization issues in this PR and making it bigger. Tracking issue [here](https://github.com/celestiaorg/rollmint/issues/551).

File: state/executor.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply state/executor.go]: Added a tracking issue [here](https://github.com/celestiaorg/rollmint/issues/552)!

File: block/manager.go (hunk 5 above)
Comment (Manav-Aggarwal) [reply block/manager.go:237]: This [issue](https://github.com/celestiaorg/rollmint/issues/549) covers it.

File: node/node.go (hunk 7 above)
Comment (Manav-Aggarwal) [reply node/node.go:364]: This [issue](https://github.com/celestiaorg/rollmint/issues/549) covers it.

File: node/integration_test.go (hunk 6 above)
Comment (tzdybal) [reply node/integration_test.go:185]: It's to give nodes some time to start and connect to each other. Value is big to ensure it's not flaky on GitHub CI.

Comment (tzdybal) [inline node/integration_test.go:185]:
```suggestion
	// wait for nodes to start up and establish connections; 1 second ensures that test pass even on CI.
	time.Sleep(1 * time.Second)
```

File: proto/tendermint/abci/types.proto (hunk 2 above)
Comment (tzdybal) [reply proto/tendermint/abci/types.proto]: After rebasing, this looks as it should :+1:

File: proto/tendermint/abci/types.proto [hunk 11]
```diff
@@ -385,6 +423,34 @@ message Snapshot {
...
+
+  RequestBeginBlock fraudulentBeginBlock = 4;
+  RequestDeliverTx fraudulentDeliverTx = 5;
+  RequestEndBlock fraudulentEndBlock = 6;
+}
+
+// State witness with a list of all witness data
+message StateWitness {
+  // store level proof
+  tendermint.crypto.ProofOp proof_op = 1;
```
[the remark this answers is not in the document]
Comment (tzdybal) [reply proto/tendermint/abci/types.proto:440]: This has to be updated there: https://github.com/celestiaorg/tendermint/pull/67, and then pulled to rollmint with `get_deps.sh`.

File: proto/tendermint/version/types.proto (hunk 3 above)
Comment (Manav-Aggarwal) [reply proto/tendermint/version/types.proto]: Removed it

File: node/integration_test.go [hunk 12]
```diff
@@ -162,12 +163,103 @@ func TestTxGossipingAndAggregation(t *testing.T) {
...
+				generateFraudProofCnt++
+			}
+		}
+		aggregatorHeight := nodes[0].Store.Height()
+		adjustedHeight := int(aggregatorHeight - 3) // 3 is completely arbitrary
+		assert.GreaterOrEqual(beginCnt, adjustedHeight)
+		assert.GreaterOrEqual(endCnt, adjustedHeight)
+		assert.GreaterOrEqual(commitCnt, adjustedHeight)
+
+		assert.Equal(generateFraudProofCnt, beginCnt+endCnt+clientNodes)
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply node/integration_test.go]: Updated and added comment with explanation.

File: node/integration_test.go [hunk 13]
```diff
@@ -162,12 +163,103 @@ func TestTxGossipingAndAggregation(t *testing.T) {
 			require.NoError(err)
 			nodeBlock, err := nodes[i].Store.LoadBlock(h)
 			require.NoError(err)
+			// Only Intermediate state roots set by block aggregator are relevant, removed for sake of comparison
```
[the remark this answers is not in the document]
Comment (Manav-Aggarwal) [reply node/integration_test.go:166]: Issue here: https://github.com/celestiaorg/rollmint/issues/566
